spanner/client_test.go

```diff
@@ -1757,6 +1759,43 @@ func TestClient_WithGRPCConnectionPoolAndNumChannels_Misconfigured(t *testing.T)
 	}
 }
 
+func TestClient_WithCallOptions(t *testing.T) {
+	co := &vkit.CallOptions{
+		CreateSession: []gax.CallOption{
+			gax.WithRetry(func() gax.Retryer {
+				return gax.OnCodes([]codes.Code{
+					codes.Unavailable, codes.DeadlineExceeded,
+				}, gax.Backoff{
+					Initial:    200 * time.Millisecond,
+					Max:        30000 * time.Millisecond,
+					Multiplier: 1.25,
+				})
+			}),
+		},
+	}
+
+	_, client, teardown := setupMockedTestServerWithConfig(t, ClientConfig{CallOptions: co})
+	defer teardown()
+
+	c, err := client.sc.nextClient()
+	if err != nil {
+		t.Fatalf("failed to get a session client: %v", err)
+	}
+
+	cs := &gax.CallSettings{}
+	// This is the default retry setting.
+	c.CallOptions.CreateSession[0].Resolve(cs)
+	if got, want := fmt.Sprintf("%v", cs.Retry()), "&{{250000000 32000000000 1.3 0} [14]}"; got != want {
+		t.Fatalf("merged CallOptions is incorrect: got %v, want %v", got, want)
+	}
+
+	// This is the custom retry setting.
+	c.CallOptions.CreateSession[1].Resolve(cs)
```

Comment:
Is this the setting that will actually be used in this case? Is it in that case necessary to keep the other settings still in `c.CallOptions.CreateSession[0]`?

Reply:
As I described in the previous comment, `c.CallOptions.CreateSession[1]` has higher order of precedence. Yes, it is not necessary to keep `c.CallOptions.CreateSession[0]`. But I have two reasons to do it in this way: 

- Follow the same pattern in `spanner_client.go`: `opts = append(c.CallOptions.CreateSession[0:len(c.CallOptions.CreateSession):len(c.CallOptions.CreateSession)], opts...)`
- Simplify the merging logic: we don't need to say `if b is not empty, use b`, `else if a is not empty, use a`, and `if both empty, leave it empty`. We can just merge `a` and `b` in the right order.

Reply:
Sounds reasonable to me. I do think we should consider adding some kind of end-to-end test for this, as we are relying on what is effectively an undocumented feature of the generated client that might silently change in the future. That end-to-end test could probably be created using the mock server, and does not need to be a full integration test.